File: dsa-learning/dsa-python-code/com/dsa/leetcode/linked_list/P142_LinkedListCycleII.py
# author: tobebetter9527
# since: 2023/10/31 11:31
from typing import Optional

from com.dsa.leetcode.linked_list import ListNode


class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        fast = slow = head
        while True:
            if not fast or not fast.next:
                return None
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                break
        fast = head
        while fast != slow:
            fast = fast.next
            slow = slow.next
        return slow

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/linked_list/P142_LinkedListCycleII.java
package com.dsa.leetcode.labuladong.linked_list;

public class P142_LinkedListCycleII {
    public ListNode detectCycle(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode fast = head, slow = head;
        while (true) {
            if (fast == null || fast.next == null) {
                return null;
            }
            fast = fast.next.next;
            slow = slow.next;
            if (slow == fast) {
                break;
            }
        }
        fast = head;
        while (fast != slow) {
            fast = fast.next;
            slow = slow.next;
        }
        return slow;
    }
}
